File: meridian-web/frontend/src/api/transcriptionAuthority.ts
import api from './client';
import type { TranscriptionAuthorityState } from '../types';

export interface AuthoritativeSegmentWire {
  segment_key: string;
  source: string;
  side: string | null;
  speaker: string | null;
  text: string;
  start_ms: number;
  end_ms: number;
}

export interface AuthoritativeTranscriptResponse {
  meeting_id: number;
  available: boolean;
  epochs_count: number;
  sources_used: string[];
  segment_count: number;
  truncated: boolean;
  segments: AuthoritativeSegmentWire[];
}

export async function getTranscriptionAuthorityState(
  meetingId: number,
): Promise<TranscriptionAuthorityState> {
  const { data } = await api.get<TranscriptionAuthorityState>(
    `/meetings/${meetingId}/transcription-authority/state`,
  );
  return data;
}

export async function promoteTranscription(
  meetingId: number, force = false,
): Promise<TranscriptionAuthorityState> {
  const { data } = await api.post<TranscriptionAuthorityState>(
    `/meetings/${meetingId}/transcription-authority/promote`, { force },
  );
  return data;
}

export async function fallbackTranscription(
  meetingId: number,
): Promise<TranscriptionAuthorityState> {
  const { data } = await api.post<TranscriptionAuthorityState>(
    `/meetings/${meetingId}/transcription-authority/fallback`, {},
  );
  return data;
}

export async function getAuthoritativeTranscript(
  meetingId: number,
): Promise<AuthoritativeTranscriptResponse> {
  const { data } = await api.get<AuthoritativeTranscriptResponse>(
    `/meetings/${meetingId}/transcription-authority/transcript`,
  );
  return data;
}

File: meridian-web/backend/app/schemas/transcription_authority.py
"""Схемы API авторитетного источника транскрипта (Этап 9.8)."""

from typing import Any

from pydantic import BaseModel


class CutoverRolloutOut(BaseModel):
    allowed: bool
    reason: str
    bucket: int


class TranscriptionAuthorityStateOut(BaseModel):
    meeting_id: int
    current_source: str
    revision: int
    fallback_used: bool
    epochs_count: int
    can_promote: bool
    rollout: CutoverRolloutOut
    quality: dict[str, Any] | None = None
    last_switch: dict[str, Any] | None = None


class PromoteRequest(BaseModel):
    reason: str | None = None
    force: bool = False


class FallbackRequest(BaseModel):
    reason: str | None = None


class AuthoritativeSegmentOut(BaseModel):
    segment_key: str
    source: str
    side: str | None = None
    speaker: str | None = None
    text: str
    start_ms: int
    end_ms: int


class AuthoritativeTranscriptOut(BaseModel):
    meeting_id: int
    available: bool          # есть ли эпохи (иначе single-only — постхок не строим)
    epochs_count: int
    sources_used: list[str]
    segment_count: int
    truncated: bool
    segments: list[AuthoritativeSegmentOut]
